Tags: backport #928 to eight

jaredgrubb opened the pull request:
This is a backport to eight of the following change:
> commit 0d222614c5996b0b3bc3ca0dc30629c4199f50a4
> Author: Jared Grubb <jaredgrubb@gmail.com>
> Date:   Mon Feb 10 12:53:43 2014 -0800
> 
>     Fix #928: Allow builder to be associated with multiple tags

For the most part, this patch is identical to the one I contributed to nine. However, the Status parts of the patch (eg, waterfall) are all brand new, since Status does not exist in this form in Nine. 

I'm sure there's going to be some issues, as I had some questions on parts of this patch (I'll write them inline below). 
Also, docs arent ported and notes are not updated. I'll do that next, but wanted to get this up for comment.

sa2ajj commented:
(_just_ 29 files... easy to review...)

jaredgrubb commented on master/buildbot/config.py:
```diff
@@ -688,10 +689,25 @@ def __init__(self, name=None, slavename=None, slavenames=None,
+                error("builder '%s': category must be a string" % (name,))
+            tags = [category]
+
+            self.category = category  # Set this for legacy reasons
```
This line was not part of the Nine review. I added it just in case subclasses had come to depend on it being there, but maybe that's not a concern.

jaredgrubb commented on master/buildbot/status/builder.py:
```diff
@@ -62,23 +62,23 @@ class BuilderStatus(styles.Versioned):
     implements(interfaces.IBuilderStatus, interfaces.IEventSource)
 
-    persistenceVersion = 1
+    persistenceVersion = 2
```
Note that this persistence is identical to the version done in Nine, so there is no version conflict (ie, both versions have a "version 2" and they are identical).

jaredgrubb commented on master/buildbot/status/web/builder.py:
```diff
@@ -602,10 +602,13 @@ def content(self, req, cxt):
+            if not showTags:
+                showTags = None
+
+        builders = req.args.get("builder", status.getBuilderNames(tags=showTags))
```
For the status objects, I decided it was probably better to not break existing bookmarks, so all the request checks will accept 'category' (if 'tag' wasnt there).

jaredgrubb commented on master/buildbot/status/web/console.py:
```diff
@@ -284,24 +284,26 @@ def getAllBuildsForRevision(self, status, request, codebase, lastRevision,
-            if not builderList.get(category):
-                builderList[category] = []
+            if len(tags)==1:
+                tags = tags[0].split('|')
```
I'm really not sure about these two lines (300, 301), because I dont fully understand what the intent was in the comment. But, I'm hoping I preserved functionality.

sa2ajj replied:
Two questions:

* do we have tests for this special approach?
* if the intent is not clear, why keep it at all?

sa2ajj replied:
Chromium uses version 0.8.4p1 (with own patches).  If they upgraded to the latest release (or, at least, to 0.8.9), I'd see some reason to keep this hack in.

sa2ajj replied:
@jaredgrubb so what about this functionality?  I'd rather see it gone...

jaredgrubb replied:
I'm ok to remove it ... I dont fully understand why it's there. Looks like it was added in the first commit of Console (looks like 0.7.10) and hasnt been updated since. So, I would guess it's probably not used anymore? Plus, I think the "multiple tags" actually implements what this is trying to accomplish anyway. Shall I remove it?

sa2ajj replied:
Yes, please :)

jaredgrubb commented on master/buildbot/status/builder.py:
```diff
@@ -322,12 +328,27 @@ def getLastFinishedBuild(self):
+        tags = self.getTags()
+        if tags:
+            return tags[0]
+        return None
```
Note that I did not remove the {get,set}Category functions here. They shouldnt really be used, but I wasnt sure if that would unncecesarily break user code. So, I figured it was easier to just leave it in and have it return the first tag.

sa2ajj replied:
I'd add a deprecation warning here and in nine these can be just removed (if they are not already).

jaredgrubb replied:
The Status objects do not exist in Nine, so there's not really any deprecation on this object.

sa2ajj replied:
Ok.

sa2ajj commented on master/buildbot/status/builder.py:
```diff
@@ -322,12 +328,27 @@ def getLastFinishedBuild(self):
+        self.tags = tags
+
+    def matchesAnyTag(self, tags):
+        return self.tags and any((tag in self.tags) for tag in tags)
```
If either parameter `tags` or `self.tags` is empty, the result is `False`, this is expected, right?

jaredgrubb replied:
builder.matchesAnyTags(["foo"]) should be False if that builder has no tags (eg, looking for the "Linux" builders should not match things that never opted in to be "Linux" builders).

builder.matchesAnyTags([]) ... I am not sure what this should be. I could make arguments both ways.

sa2ajj replied:
Let's keep it the way you already have it then.

sa2ajj commented on master/buildbot/status/builder.py:
```diff
@@ -322,12 +328,27 @@ def getLastFinishedBuild(self):
+        return self.tags and any((tag in self.tags) for tag in tags)
+
+    def matchesAllTags(self, tags):
+        return self.tags and all((tag in self.tags) for tag in tags)
```
Side note: using sets would be faster here and in the previous function.

sa2ajj commented on master/buildbot/status/mail.py:
```diff
@@ -299,12 +301,15 @@ def __init__(self, fromaddr, mode=("failing", "passing", "warnings"),
                            but not both.
 
+        @type  categories: list of strings
+        @param categories: (this attribute is deprecated; use 'tags' instead)
```
No deprecation warning in this constructor?

sa2ajj commented on master/buildbot/status/web/console.py:
```diff
@@ -316,26 +318,26 @@ def getAllBuildsForRevision(self, status, request, codebase, lastRevision,
         cs = []
 
-        for category in categories:
+        for tag in tags:
```
[nit] This loop calls for a list comprehension...

jaredgrubb replied:
Yes, this would definitely look better as a list-comp ... In general, I tried to not rewrite code that wasnt actually changing behavior. I just have been wrong too many times by "obviously the same thing" kinds of changes :) Since this code is dead-end in eight, I'm tempted to just let it be unless you have a strong objection.

sa2ajj replied:
No strong objection here :)

sa2ajj commented:
LGTM

sa2ajj commented:
pyflakes error:

```
master/buildbot/status/web/waterfall.py:325: import 'builder' from line 32 shadowed by loop variable
```

sa2ajj commented:
@jaredgrubb do you think it will be mergeable by this weekend?

jaredgrubb commented:
Ok, I think I addressed all the comments. Also, I have updated the docs and release notes. (Note that when I did 'make docs', it failed, but it seemed to fail for other reasons).

sa2ajj commented:
Looks like a pep-8 error:

```
master/buildbot/status/tinderbox.py:69:33: E261 at least two spaces before inline comment
categories=None # deprecated, use tags instead
^

master/buildbot/status/web/console.py:294:1: W293 blank line contains whitespace
^

master/buildbot/status/web/console.py:300:25: E225 missing whitespace around operator
if len(tags)==1:
^

master/buildbot/test/unit/test_status_mail.py:189:5: E303 too many blank lines (2)
def test_init_enforces_categories_and_builders_are_mutually_exclusive(self):
^

1 E225 missing whitespace around operator
1 E261 at least two spaces before inline comment
1 E303 too many blank lines (2)
1 W293 blank line contains whitespace
```

jaredgrubb commented:
Fixed two of the pep8 issues (the other two were removed in that chromium patch).

sa2ajj commented:
Thank you very much! :)

Waiting for Travis...

sa2ajj commented:
We seem to have an error: http://buildbot.buildbot.net/console

sa2ajj commented:
[TRAC-3179](http://trac.buildbot.net/ticket/3179) has a better traceback.

sa2ajj commented on master/buildbot/status/web/console.py:
```diff
@@ -404,21 +397,21 @@ def displayStatusLine(self, builderList, allBuilds, revision, debugInfo):
-        # Display the boxes by category group.
-        for category in categories:
+        # Display the boxes by tag group.
+        for tag in tags:
```
Inside this `for..` loop there was a variable `tag` used, which overwrites the value assigned here.

sa2ajj commented on master/buildbot/status/web/console.py:
```diff
@@ -475,7 +468,7 @@ def displayStatusLine(self, builderList, allBuilds, revision, debugInfo):
                 b["color"] = resultsClass
                 b["tag"] = tag
```
This tag has nothing to do with the `tag` assigned in the loop.

sa2ajj commented:
#1514 partially fixes things.
